STM: Update linker script for using SRAM1 and SRAM2 in ARM

deepikabhavnani opened the pull request:
### Description

To have the flexibilty in application; to use any of the section (data/bss/heap) without updating linker script in every use case, following decisions are made:

1. Fixed size and small sections moved to SRAM2 (32K)
    Vectors
    Crash data
    Stack
    Remaining section - RW / ZI
2. Large memory space should be used for variable sections
   RW/ZI
   Heap - (Minimum - xxx target based) 

### Pull request type
    [X] Fix
    [ ] Refactor
    [ ] Target update
    [ ] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

jeromecoutant commented on TESTS/mbedmicro-rtos-mbed/heap_and_stack/main.cpp:
```diff
@@ -204,8 +204,9 @@ void test_isr_stack_in_range(void)
-    bool result = inrange(msp, mbed_stack_isr_start + mbed_stack_isr_size - 128, 128);
-
+    printf ("mbed_stack_isr_start = 0x%x mbed_stack_isr_size = 0x%x\n", mbed_stack_isr_start, mbed_stack_isr_size);
+    bool result = inrange(msp, mbed_stack_isr_start, mbed_stack_isr_size);
```
I don't understand why a PR tagged as STM needs to update tests?

deepikabhavnani replied:
Ohh this change in test should already be on master now.

jeromecoutant commented on targets/TARGET_STM/TARGET_STM32L4/TARGET_STM32L443xC/device/TOOLCHAIN_ARM_STD/stm32l443xx.sct:
```diff
@@ -29,36 +29,82 @@
+  #define MBED_RAM_SIZE             0xC000
+#endif
+
+#if !defined(MBED_RAM2_START)
```
This macro comes from CMSIS pack ?
So is it mandatory to always define a default value ?

deepikabhavnani replied:
Yes because the macro is yet not available for all targets.

jeromecoutant commented on targets/TARGET_STM/mbed_rtx.h:
```diff
@@ -137,4 +137,13 @@
+#if defined(__ARMCC_VERSION)
+    extern uint32_t               Image$$ARM_LIB_HEAP$$ZI$$Base[];
+    extern uint32_t               Image$$ARM_LIB_HEAP$$ZI$$Length[];
+    #define HEAP_START            Image$$ARM_LIB_HEAP$$ZI$$Base
```
There used to have this kind of of definition depending on toolchain.
But it has been removed some time ago.
So I don't think adding this back depending on toolchain and targets is a so good thing?

deepikabhavnani replied:
This definition is needed to over write the defaults used in retarget and toolchain specific file. Earlier the changes were present for 2-region memory model for specific target,  with change in memory model of Mbed os this was not needed and hence removed. This addition also indicates that the linker script here does not follow the default memory model, which is stack at the bottom of RAM memory

jeromecoutant reviewed (commented):
Few questions.
But as a general question, not easy to approve a specific PR for 2 targets and 1 toolchain only
@LMESTM

Commit:
Update linker script for using SRAM1 and SRAM2 in ARM
    To have the flexibilty in application; to use any of the section
    (data/bss/heap) without updating linker script in every use case,
    following decisions are made:
    1. Fixed size and small sections moved to SRAM2 (32K)
        Vectors
        Crash data
        Remaining section - RW / ZI
    2. Large memory space should be used for variable sections
       RW/ZI
       Heap - (Minimum - 0x12000)
       Stack - At bottom

jeromecoutant reviewed (commented):
I could approve, but who takes the action to align all scatter files of all targets...

Commit:
Add missing boot stack size memory from heap calculation
